### PR TITLE
support to merge complex commands when to inject commands

### LINKED ISSUE #143: Cannot start tiflash when to inject iochaos
## Bug Report

**What version of Kubernetes are you using?**

```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:54:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:43:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
```
**What did you do?**
inject iochaos 

tiflash configmap 
```
apiVersion: v1
data:
  chaosfs-tiflash.yaml: |
    name: chaosfs-tiflash
    selector:
      labelSelectors:
        "app": "tiflash"
    initContainers:
    - name: inject-scripts
      image: pingcap/chaos-scripts:latest
      imagePullpolicy: Always
      command: ["sh", "-c", "/scripts/init.sh -d /data/data/db -f /data/data/fuse-data"]
    containers:
    - name: chaosfs
      image: pingcap/chaos-fs:latest
      imagePullpolicy: Always
      ports:
      - containerPort: 65534
      securityContext:
        privileged: true
      command:
        - /usr/local/bin/chaosfs
        - -addr=:65534
        - -pidfile=/tmp/fuse/pid
        - -original=/data/data/fuse-data
        - -mountpoint=/data/data/db
      volumeMounts:
        - name: tiflash
          mountPath: /data/data
          mountPropagation: Bidirectional
    volumeMounts:
    - name: tiflash
      mountPath: /data/data
      mountPropagation: HostToContainer
    - name: scripts
      mountPath: /tmp/scripts
    - name: fuse
      mountPath: /tmp/fuse
    volumes:
    - name: scripts
      emptyDir: {}
    - name: fuse
      emptyDir: {}
    postStart:
      tiflash:
        command:
          - /tmp/scripts/wait-fuse.sh
``` 
error log 
```
E0117 13:35:00.855365       1 stateful_set.go:400] Error syncing StatefulSet tiflash-test5/tiflash, requeuing: Internal error occurred: Internal error occurred: jsonpatch add operation does not apply: doc is missing path: "/
metadata/annotations/admission-webhook.pingcap.com~1status"
I0117 13:35:00.856044       1 event.go:221] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"tiflash-test5", Name:"tiflash", UID:"3956b26b-385b-11ea-b391-4cd98f4bd3ae", APIVersion:"apps/v1", ResourceVersion:"393890116
", FieldPath:""}): type: 'Warning' reason: 'FailedCreate' create Pod tiflash-1 in StatefulSet tiflash failed error: Internal error occurred: Internal error occurred: jsonpatch add operation does not apply: doc is missing pat
h: "/metadata/annotations/admission-webhook.pingcap.com~1status"
E0117 13:35:00.868822       1 stateful_set.go:400] Error syncing StatefulSet tiflash-test5/tiflash, requeuing: Internal error occurred: Internal error occurred: jsonpatch add operation does not apply: doc is missing path: "/
metadata/annotations/admission-webhook.pingcap.com~1status"
I0117 13:35:00.868891       1 event.go:221] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"tiflash-test5", Name:"tiflash", UID:"3956b26b-385b-11ea-b391-4cd98f4bd3ae", APIVersion:"apps/v1", ResourceVersion:"393890116
", FieldPath:""}): type: 'Warning' reason: 'FailedCreate' create Pod tiflash-1 in StatefulSet tiflash failed error: Internal error occurred: Internal error occurred: jsonpatch add operation does not apply: doc is missing pat
h: "/metadata/annotations/admission-webhook.pingcap.com~1status"
E0117 13:35:00.886926       1 stateful_set.go:400] Error syncing StatefulSet tiflash-test5/tiflash, requeuing: Internal error occurred: Internal error occurred: jsonpatch add operation does not apply: doc is missing path: "/
metadata/annotations/admission-webhook.pingcap.com~1status"
```

### PR DESCRIPTION
**Pull request by cwen0**
Signed-off-by: cwen0 <cwenyin0@gmail.com>

### What problem does this PR solve?  

fix #143  

### What is changed and how does it work? 

1. support to merge complex commands
2. fix update annotations when to inject the sidecar container

eg: 
```
inject: []string{"bash", "-c", "/check.sh"}, origin: []string{"bash", "-c", "set -ex\n/tiflash server --config-file /data/config.toml"}
merged commands: []string{"/bin/sh", "-ec", "/check.sh\nset -ex\n/tiflash server --config-file /data/config.toml\n"}
``` 

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

image

log:  

```
kubectl logs -f --namespace app-ns tiflash-0 -c tiflash
fuse server not running, /tmp/fuse/pid not found, retry...
fuse server not running, /tmp/fuse/pid not found, retry...
fuse server not running, /tmp/fuse/pid not found, retry...
fuse server not running, /tmp/fuse/pid not found, retry...
fuse server not running, /tmp/fuse/pid not found, retry...
fuse server is started
++ hostname
+ [[ tiflash-0 =~ -([0-9]+)$ ]]
+ ordinal=0
+ set +ex
Server:         10.96.0.10
Address:        10.96.0.10#53

Name:   tiflash-0.tiflash.app-ns.svc.cluster.local
Address: 10.244.5.18

Logging trace to /logs/server.log
Logging errors to /logs/error.log
```

Code changes

 - Has Go code change

### CODE REVIEW

File: pkg/utils/command.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+func mergeOriginCommandsAndArgs(origin []string, args []string) string {
+	commands := []string{}
+	commands = append(commands, origin...)
+	commands = append(commands, args...)
+
+	return mergeCommandsAction(commands)
+}
+
+func mergeCommandsAction(commands []string) string {
```
Comment (zhouqiang-cl) [inline pkg/utils/command.go]: Can we split this function, it too long

File: pkg/utils/command.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	// merge injected commands
+	scripts := mergeCommandsAction(inject)
+
+	// merge original commands
+	scripts += mergeOriginCommandsAndArgs(origin, args)
+
+	return []string{"/bin/sh", "-ec", scripts}
+}
+
+func mergeOriginCommandsAndArgs(origin []string, args []string) string {
```
Comment (zhouqiang-cl) [inline pkg/utils/command.go]: mv mergeOriginCommandsAndArgs after mergeCommandsAction?

File: pkg/utils/command.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+		scripts += cmd
+		scripts += "\n"
+		continue
+	}
+
+	return scripts
+}
+
+func isCommonScripts(cmd string) bool {
```
Comment (zhouqiang-cl) [inline pkg/utils/command.go]: split to few function? how about isShellScript, isPythonScript, and isCommonScript check if it is shell or python

File: pkg/webhook/inject/inject.go [hunk 4]
```diff
@@ -520,19 +507,19 @@ func mergeVolumeMounts(volumeMounts []corev1.VolumeMount, containers []corev1.Co
 
 func updateAnnotations(target map[string]string, added map[string]string) (patch []patchOperation) {
 	for key, value := range added {
-		keyEscaped := strings.Replace(key, "/", "~1", -1)
-
 		if target == nil || target[key] == "" {
 			target = map[string]string{}
 			patch = append(patch, patchOperation{
-				Op:    "add",
-				Path:  "/metadata/annotations/" + keyEscaped,
```
Comment (zhouqiang-cl) [inline pkg/webhook/inject/inject.go]: why not need keyEscaped

File: pkg/utils/command_test.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+			},
+			origin: []string{
+				"bash",
+				"-c",
+				"set -ex\n[[ `hostname` =~ -([0-9]+)$ ]] || exit 1\n/tiflash server --config-file /data/config.toml",
+			},
+			expectedValue: []string{
+				"/bin/sh",
+				"-ec",
+				"/check.sh -v 1\nset -ex\n[[ `hostname` =~ -([0-9]+)$ ]] || exit 1\n/tiflash server --config-file /data/config.toml\n",
```
Comment (zhouqiang-cl) [inline pkg/utils/command_test.go]: Did check.sh handle \n

File: pkg/utils/command.go (hunk 2 above)
Comment (cwen0) [reply pkg/utils/command.go]: why?

Comment (zhouqiang-cl) [reply pkg/utils/command.go]: MergeCommands first call mergeCommandsAction, then call mergeOriginCommandsAndArgs

File: pkg/utils/command.go (hunk 1 above)
Comment (cwen0) [reply pkg/utils/command.go]: This is a relatively complete logic. If it is broken into small functions, the function name is not good.

File: pkg/webhook/inject/inject.go (hunk 4 above)
Comment (cwen0) [reply pkg/webhook/inject/inject.go]: use `k-v` type to set annotations, so `keyEscaped` is unnecessary

File: pkg/utils/command_test.go (hunk 5 above)
Comment (cwen0) [reply pkg/utils/command_test.go]: No, keep `\n`

File: pkg/utils/command.go (hunk 3 above)
Comment (cwen0) [reply pkg/utils/command.go]: Great, I will update it.

File: pkg/utils/command.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+				}
+			}
+			scripts += tempScripts
+			scripts += "\n"
+			break
+		}
+
+		scripts += cmd
+		scripts += "\n"
+		continue
```
Comment (mahjonp) [inline pkg/utils/command.go]: emm, this continue is not necessary

**Comment by mahjonp** [comment]
seems this PR is workable, but I'm not sure. @you06 @mapleFU PTAL

File: pkg/utils/command.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	if isShellScripts(cmd) || isPythonScripts(cmd) {
+		return true
+	}
+
+	return false
+}
+
+func isShellScripts(cmd string) bool {
+	if cmd == "bash" || cmd == "sh" ||
+		strings.HasPrefix(cmd, "bash ") || strings.HasPrefix(cmd, "sh ") ||
```
Comment (mapleFU) [inline pkg/utils/command.go]: Why not just `strings.HasPrefix(cmd, "bash ") || strings.HasPrefix(cmd, "sh ")`

File: pkg/utils/command.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+		scripts += cmd
+		scripts += "\n"
+	}
+
+	return scripts
+}
+
+func mergeOriginCommandsAndArgs(origin []string, args []string) string {
+	commands := []string{}
```
Comment (mapleFU) [inline pkg/utils/command.go]: `commands := append(origin, args...)`

**Comment by mahjonp** [comment]
please comments on documents? list the limitation of current implementation, Rest LGTM

File: pkg/utils/command.go (hunk 7 above)
Comment (cwen0) [reply pkg/utils/command.go]: avoid some commands with prefix `bash` or `sh`, eg: `shell2 xxx`

**Comment by cwen0** [comment]
@mahjonp @zhouqiang-cl @YangKeao @mapleFU PTAL again, thanks!

File: doc/io_chaos.md [hunk 9]
```diff
@@ -10,6 +10,12 @@ read/write errors, etc. It can inject delay and errno when you use the IO system
 
 ## Prerequisites
 
+### Commands and args for the application container
+
+Chaos Mesh uses [`wait-fush.sh`](https://github.com/pingcap/chaos-mesh/blob/master/doc/sidecar_configmap.md#tips) to ensure that the fuse-daemon server is running normally before the application starts. 
+so `wait-fush.sh` needs to be injected into the startup command of the container. If the application process is not stared by the [commands and args of the container](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/),
```
Comment (mahjonp) [inline doc/io_chaos.md]:
```suggestion
so `wait-fush.sh` needs to be injected into the startup command of the container. If the application process is not started by the [commands and args of the container](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/), 
```

**Pull request merged**